mlflow/openai/api_request_parallel_processor.py

```
@@ -71,279 +59,55 @@ def increment_num_rate_limit_errors(self):
-                    available_token_capacity -= next_request_tokens
-                    next_request.attempts_left -= 1
-                    # call API
-                    api_token.refresh(logger=_logger)
-                    executor.submit(
-                        next_request.call_api,
-                        retry_queue=retry_queue,
-                        status_tracker=status_tracker,
-                        headers={**api_token.auth_headers(), **extra_headers},
-                    )
-                    next_request = None  # reset next_request to empty
-                else:
-                    next_request = None
-                    status_tracker.complete_task(success=False)
-                    status_tracker.error = mlflow.MlflowException(
-                        "Request size exceeded max tokens."
-                    )
-
-            # if all tasks are finished, break
-            if requests_exhausted and status_tracker.num_tasks_in_progress == 0:
-                break
-
-            # if a rate limit error was hit recently, pause to cool down
-            seconds_since_rate_limit_error = (
-                time.time() - status_tracker.time_of_last_rate_limit_error
-            )
-            if seconds_since_rate_limit_error < seconds_to_pause_after_rate_limit_error:
-                remaining_seconds_to_pause = (
-                    seconds_to_pause_after_rate_limit_error - seconds_since_rate_limit_error
-                )
-                _logger.debug(
-                    "Encountered rate limit error. Pausing to cool down for "
-                    f"{remaining_seconds_to_pause} seconds..."
-                )
-                time.sleep(remaining_seconds_to_pause)
-                # ^e.g., if pause is 15 seconds and final limit was hit 5 seconds ago
-
-            time.sleep(0.001)  # avoid busy waiting
+            for index, task in requests_iter
+        ]
+        wait(futures, return_when=ALL_COMPLETED)
```

```suggestion
        wait(futures, return_when=FIRST_EXCEPTION)
```

should we use `FIRST_EXCEPTION` to fail fast?

---

I just tried to keep the existing logic here, the current logic let all tasks run and count the number of failures. I personally agree with fail fast.